File: docx_parsers/numbering_parser.py
# numbering_parser.py

from pydantic import BaseModel
from typing import List, Optional
from docx_parsers.utils import extract_xml_root_from_docx, read_binary_from_file_path, convert_twips_to_points
from docx_parsers.helpers.common_helpers import extract_element, extract_attribute, NAMESPACE
from docx_parsers.models.styles_models import FontProperties, IndentationProperties
import json

class NumberingLevel(BaseModel):
    numId: int
    ilvl: int
    start: int
    numFmt: str
    lvlText: str
    lvlJc: str
    counter: int = None
    indent: Optional[IndentationProperties] = None
    tab_pt: Optional[float] = None
    fonts: Optional[FontProperties] = None

class NumberingInstance(BaseModel):
    numId: int
    levels: List[NumberingLevel]

class NumberingSchema(BaseModel):
    instances: List[NumberingInstance]

class NumberingParser:
    def __init__(self, docx_file: bytes):
        self.root = extract_xml_root_from_docx(docx_file, 'numbering.xml')
        self.numbering_schema = self.parse()

    def parse(self) -> NumberingSchema:
        instances = []
        
        for num in self.root.findall(".//w:num", namespaces=NAMESPACE):
            numId = int(extract_attribute(num, 'numId'))
            abstractNumId = int(extract_attribute(extract_element(num, ".//w:abstractNumId"), 'val'))
            levels = []

            abstractNum = extract_element(self.root, f".//w:abstractNum[@w:abstractNumId='{abstractNumId}']")
            for lvl in abstractNum.findall(".//w:lvl", namespaces=NAMESPACE):
                ilvl = int(extract_attribute(lvl, 'ilvl'))
                start = int(extract_attribute(extract_element(lvl, ".//w:start"), 'val'))
                numFmt = extract_attribute(extract_element(lvl, ".//w:numFmt"), 'val')
                lvlText = extract_attribute(extract_element(lvl, ".//w:lvlText"), 'val')
                lvlJc = extract_attribute(extract_element(lvl, ".//w:lvlJc"), 'val')
                
                pPr = extract_element(lvl, ".//w:pPr")
                indent_properties = IndentationProperties()
                tab_pt = None

                if pPr is not None:
                    indent_element = extract_element(pPr, "w:ind")
                    if indent_element is not None:
                        indent_properties.left_pt = convert_twips_to_points(int(extract_attribute(indent_element, 'left') or extract_attribute(indent_element, 'start') or 0))
                        indent_properties.right_pt = convert_twips_to_points(int(extract_attribute(indent_element, 'right') or extract_attribute(indent_element, 'end') or 0))
                        indent_properties.hanging_pt = convert_twips_to_points(int(extract_attribute(indent_element, 'hanging') or 0))
                        indent_properties.firstline_pt = convert_twips_to_points(int(extract_attribute(indent_element, 'firstLine') or 0))
                    
                    tab_element = extract_element(pPr, ".//w:tab")
                    if tab_element is not None:
                        tab_val = extract_attribute(tab_element, 'pos')
                        if tab_val.isdigit():
                            tab_pt = convert_twips_to_points(int(tab_val))
                
                # Extract font properties
                fonts = None
                rPr = extract_element(lvl, ".//w:rPr")
                if rPr is not None:
                    rFonts = extract_element(rPr, "w:rFonts")
                    if rFonts is not None:
                        fonts = FontProperties(
                            ascii=extract_attribute(rFonts, 'ascii'),
                            hAnsi=extract_attribute(rFonts, 'hAnsi'),
                            eastAsia=extract_attribute(rFonts, 'eastAsia'),
                            cs=extract_attribute(rFonts, 'cs')
                        )

                level = NumberingLevel(
                    numId=numId, ilvl=ilvl, start=start, numFmt=numFmt, lvlText=lvlText, lvlJc=lvlJc,
                    indent=indent_properties, tab_pt=tab_pt, fonts=fonts
                )
                levels.append(level)
            
            instance = NumberingInstance(numId=numId, levels=levels)
            instances.append(instance)
        
        return NumberingSchema(instances=instances)

    def get_numbering_schema(self) -> NumberingSchema:
        return self.numbering_schema

if __name__ == "__main__":

    docx_path = "C:/Users/omerh/Desktop/new_docx.docx"
    # docx_path = "C:/Users/omerh/Desktop/file-sample_1MB.docx"

    docx_file = read_binary_from_file_path(docx_path)

    numbering_parser = NumberingParser(docx_file)
    # numbering_scheme = numbering_parser.get_numbering_schema()

    filtered_schema_dict = numbering_parser.numbering_schema.model_dump(exclude_none=True)
    # Output or further process the numbering_scheme as needed
    print(json.dumps(filtered_schema_dict, indent=2))

File: docx-parser-converter-js/src/parsers/numbering/numbering-parser.ts
/**
 * Numbering parser for DOCX documents
 * Extracts and processes numbering definitions from numbering.xml
 */

import {
  type NumberingSchema,
  NumberingSchemaModel,
  type NumberingLevel,
  NumberingLevelModel,
  type NumberingInstance,
  NumberingInstanceModel,
} from '@/models/numbering-models.js';
import {
  type FontProperties,
  FontPropertiesModel,
  type IndentationProperties,
  IndentationPropertiesModel,
} from '@/models/styles-models.js';
import { convertTwipsToPoints } from '@/utils/unit-conversion.js';
import { BaseParser } from '../base-parser.js';

/**
 * Numbering parser for DOCX numbering.xml
 */
export class NumberingParser extends BaseParser<NumberingSchema> {
  constructor(options: Record<string, unknown> = {}) {
    super('NumberingParser', options);
  }

  /**
   * Parse numbering.xml content
   * @param xmlObj - Parsed XML object from numbering.xml
   * @returns Promise resolving to NumberingSchema
   */
  protected async parseInternal(xmlObj: Record<string, unknown>): Promise<NumberingSchema> {
    const numberingElement = this.extractNumberingElement(xmlObj);
    
    // Parse abstract numbering definitions
    const levels = await this.parseAbstractNumbering(numberingElement);
    
    // Parse numbering instances
    const instances = await this.parseNumberingInstances(numberingElement);
    
    return NumberingSchemaModel.create({
      levels,
      instances,
    });
  }

  /**
   * Extract numbering element from XML
   * @param xmlObj - Parsed XML object
   * @returns Numbering element
   */
  private extractNumberingElement(xmlObj: Record<string, unknown>): Record<string, unknown> {
    if (xmlObj['w:numbering']) {
      const numberingEl = xmlObj['w:numbering'];
      return Array.isArray(numberingEl) ? numberingEl[0] : (numberingEl as Record<string, unknown>);
    }
    
    // If no w:numbering wrapper, assume the root is the numbering element
    return xmlObj;
  }

  /**
   * Parse abstract numbering definitions
   * @param numberingElement - w:numbering element
   * @returns Array of numbering levels
   */
  private async parseAbstractNumbering(
    numberingElement: Record<string, unknown>
  ): Promise<NumberingLevel[]> {
    const levels: NumberingLevel[] = [];
    
    // Get all w:abstractNum elements
    const abstractNums = this.getChildElements(numberingElement, 'w:abstractNum');
    
    for (const abstractNum of abstractNums) {
      const abstractNumId = this.getNumberAttribute(abstractNum, 'w:abstractNumId', 0);
      
      // Parse levels within this abstract numbering
      const lvlElements = this.getChildElements(abstractNum, 'w:lvl');
      
      for (const lvlElement of lvlElements) {
        try {
          const level = await this.parseNumberingLevel(lvlElement, abstractNumId);
          levels.push(level);
        } catch (error) {
          this.addWarning(
            `Failed to parse numbering level in abstractNum ${abstractNumId}: ${
              error instanceof Error ? error.message : 'Unknown error'
            }`
          );
        }
      }
    }
    
    return levels;
  }

  /**
   * Parse a single numbering level
   * @param lvlElement - w:lvl element
   * @param numId - The numbering ID
   * @returns Parsed NumberingLevel
   */
  private async parseNumberingLevel(
    lvlElement: Record<string, unknown>,
    numId: number
  ): Promise<NumberingLevel> {
    // Extract level attributes
    const ilvl = this.getNumberAttribute(lvlElement, 'w:ilvl', 0);
    
    // Parse start value
    const startElement = this.getFirstChild(lvlElement, 'w:start');
    const start = startElement ? this.getNumberAttribute(startElement, 'w:val', 1) : 1;
    
    // Parse number format
    const numFmtElement = this.getFirstChild(lvlElement, 'w:numFmt');
    const numFmt = numFmtElement ? this.getAttribute(numFmtElement, 'w:val') || 'decimal' : 'decimal';
    
    // Parse level text
    const lvlTextElement = this.getFirstChild(lvlElement, 'w:lvlText');
    const lvlText = lvlTextElement ? this.getAttribute(lvlTextElement, 'w:val') || '' : '';
    
    // Parse justification
    const lvlJcElement = this.getFirstChild(lvlElement, 'w:lvlJc');
    const lvlJc = lvlJcElement ? this.getAttribute(lvlJcElement, 'w:val') || 'left' : 'left';
    
    // Parse paragraph properties for indentation
    const indent = await this.parseIndentationFromPPr(lvlElement);
    
    // Parse tab position
    const tab_pt = await this.parseTabPosition(lvlElement);
    
    // Parse run properties for fonts
    const fonts = await this.parseFontsFromRPr(lvlElement);

    return NumberingLevelModel.create({
      numId,
      ilvl,
      start,
      numFmt,
      lvlText,
      lvlJc,
      counter: start, // Initialize counter with start value
      indent,
      tab_pt,
      fonts,
    });
  }

  /**
   * Parse indentation from paragraph properties
   * @param lvlElement - w:lvl element
   * @returns Indentation properties or undefined
   */
  private async parseIndentationFromPPr(
    lvlElement: Record<string, unknown>
  ): Promise<IndentationProperties | undefined> {
    const pPr = this.getFirstChild(lvlElement, 'w:pPr');
    if (!pPr) return undefined;
    
    const ind = this.getFirstChild(pPr, 'w:ind');
    if (!ind) return undefined;
    
    const left = this.getAttribute(ind, 'w:left');
    const right = this.getAttribute(ind, 'w:right');
    const firstLine = this.getAttribute(ind, 'w:firstLine');
    const hanging = this.getAttribute(ind, 'w:hanging');
    
    return IndentationPropertiesModel.create({
      left_pt: left ? convertTwipsToPoints(parseInt(left)) : undefined,
      right_pt: right ? convertTwipsToPoints(parseInt(right)) : undefined,
      firstline_pt: firstLine 
        ? convertTwipsToPoints(parseInt(firstLine))
        : hanging 
        ? -convertTwipsToPoints(parseInt(hanging))
        : undefined,
    });
  }

  /**
   * Parse tab position from paragraph properties
   * @param lvlElement - w:lvl element
   * @returns Tab position in points or undefined
   */
  private async parseTabPosition(lvlElement: Record<string, unknown>): Promise<number | undefined> {
    const pPr = this.getFirstChild(lvlElement, 'w:pPr');
    if (!pPr) return undefined;
    
    const tabs = this.getFirstChild(pPr, 'w:tabs');
    if (!tabs) return undefined;
    
    const tab = this.getFirstChild(tabs, 'w:tab');
    if (!tab) return undefined;
    
    const pos = this.getAttribute(tab, 'w:pos');
    return pos ? convertTwipsToPoints(parseInt(pos)) : undefined;
  }

  /**
   * Parse fonts from run properties
   * @param lvlElement - w:lvl element
   * @returns Font properties or undefined
   */
  private async parseFontsFromRPr(
    lvlElement: Record<string, unknown>
  ): Promise<FontProperties | undefined> {
    const rPr = this.getFirstChild(lvlElement, 'w:rPr');
    if (!rPr) return undefined;
    
    const rFonts = this.getFirstChild(rPr, 'w:rFonts');
    if (!rFonts) return undefined;
    
    return FontPropertiesModel.create({
      ascii: this.getAttribute(rFonts, 'w:ascii'),
      hAnsi: this.getAttribute(rFonts, 'w:hAnsi'),
      eastAsia: this.getAttribute(rFonts, 'w:eastAsia'),
      cs: this.getAttribute(rFonts, 'w:cs'),
    });
  }

  /**
   * Parse numbering instances
   * @param numberingElement - w:numbering element
   * @returns Array of numbering instances
   */
  private async parseNumberingInstances(
    numberingElement: Record<string, unknown>
  ): Promise<NumberingInstance[]> {
    const instances: NumberingInstance[] = [];
    
    // Get all w:num elements
    const numElements = this.getChildElements(numberingElement, 'w:num');
    
    for (const numElement of numElements) {
      try {
        const numId = this.getNumberAttribute(numElement, 'w:numId', 0);
        
        const abstractNumIdElement = this.getFirstChild(numElement, 'w:abstractNumId');
        const abstractNumId = abstractNumIdElement 
          ? this.getNumberAttribute(abstractNumIdElement, 'w:val', 0)
          : 0;
        
        instances.push(NumberingInstanceModel.create({
          numId,
          abstractNumId,
        }));
      } catch (error) {
        this.addWarning(
          `Failed to parse numbering instance: ${
            error instanceof Error ? error.message : 'Unknown error'
          }`
        );
      }
    }
    
    return instances;
  }

  /**
   * Get numbering level by ID and level
   * @param numberingSchema - Parsed numbering schema
   * @param numId - Numbering ID
   * @param ilvl - Indent level
   * @returns Numbering level or undefined if not found
   */
  public static getNumberingLevel(
    numberingSchema: NumberingSchema,
    numId: number,
    ilvl: number
  ): NumberingLevel | undefined {
    return numberingSchema.levels.find(
      level => level.numId === numId && level.ilvl === ilvl
    );
  }

  /**
   * Get all levels for a numbering ID
   * @param numberingSchema - Parsed numbering schema
   * @param numId - Numbering ID
   * @returns Array of numbering levels
   */
  public static getNumberingLevels(
    numberingSchema: NumberingSchema,
    numId: number
  ): NumberingLevel[] {
    return numberingSchema.levels.filter(level => level.numId === numId);
  }

  /**
   * Get numbering instance by ID
   * @param numberingSchema - Parsed numbering schema
   * @param numId - Numbering ID
   * @returns Numbering instance or undefined if not found
   */
  public static getNumberingInstance(
    numberingSchema: NumberingSchema,
    numId: number
  ): NumberingInstance | undefined {
    return numberingSchema.instances.find(instance => instance.numId === numId);
  }

  /**
   * Check if numbering format is a bullet format
   * @param numFmt - Number format
   * @returns True if bullet format
   */
  public static isBulletFormat(numFmt: string): boolean {
    return numFmt === 'bullet';
  }

  /**
   * Generate display text for a numbering level
   * @param level - Numbering level
   * @param counter - Current counter value
   * @returns Generated display text
   */
  public static generateDisplayText(level: NumberingLevel, counter: number): string {
    const { numFmt, lvlText } = level;
    
    // Handle different number formats
    let numberText: string;
    
    switch (numFmt) {
      case 'decimal':
        numberText = counter.toString();
        break;
      case 'upperRoman':
        numberText = this.toRoman(counter).toUpperCase();
        break;
      case 'lowerRoman':
        numberText = this.toRoman(counter).toLowerCase();
        break;
      case 'upperLetter':
        numberText = this.toAlpha(counter).toUpperCase();
        break;
      case 'lowerLetter':
        numberText = this.toAlpha(counter).toLowerCase();
        break;
      case 'bullet':
        return '•'; // Default bullet character
      default:
        numberText = counter.toString();
    }
    
    // Replace placeholders in level text
    return lvlText.replace(/%(\d+)/g, (match, levelNum) => {
      const levelIndex = parseInt(levelNum) - 1;
      if (levelIndex === level.ilvl) {
        return numberText;
      }
      // For multi-level numbering, you'd need to track other levels
      return match;
    });
  }

  /**
   * Convert number to Roman numerals
   * @param num - Number to convert
   * @returns Roman numeral string
   */
  private static toRoman(num: number): string {
    const values = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];
    const symbols = ['m', 'cm', 'd', 'cd', 'c', 'xc', 'l', 'xl', 'x', 'ix', 'v', 'iv', 'i'];
    
    let result = '';
    for (let i = 0; i < values.length; i++) {
      const value = values[i];
      const symbol = symbols[i];
      if (value && symbol) {
        while (num >= value) {
          result += symbol;
          num -= value;
        }
      }
    }
    return result;
  }

  /**
   * Convert number to alphabetic representation
   * @param num - Number to convert (1-based)
   * @returns Alphabetic string
   */
  private static toAlpha(num: number): string {
    let result = '';
    while (num > 0) {
      num--; // Convert to 0-based
      result = String.fromCharCode(97 + (num % 26)) + result;
      num = Math.floor(num / 26);
    }
    return result || 'a';
  }
} 